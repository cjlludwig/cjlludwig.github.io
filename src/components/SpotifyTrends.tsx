import { useEffect, useState } from 'react';
import { getTopTracks } from "../services/spotifyService"
import { Track } from '../models/tracks';
import { Card, ProgressBar } from 'react-bootstrap';

import Carousel from 'react-multi-carousel';

import 'react-multi-carousel/lib/styles.css';
import 'bootstrap/dist/css/bootstrap.min.css';
import '../static/styles/components/SpotifyTrends.css';

const responsive = {
  desktop: {
    breakpoint: { max: 3000, min: 1024 },
    items: 5,
    slidesToSlide: 5 // optional, default to 1.
  },
  tablet: {
    breakpoint: { max: 1024, min: 464 },
    items: 2,
    slidesToSlide: 1 // optional, default to 1.
  },
  mobile: {
    breakpoint: { max: 464, min: 0 },
    items: 1,
    slidesToSlide: 1 // optional, default to 1.
  }
};

function determineTrendiness(popularity: number) {
  if (popularity <= 10) return "10";
  if (popularity <= 20) return "20"; 
  if (popularity <= 30) return "30";
  if (popularity <= 40) return "40"; 
  if (popularity <= 50) return "50"; 
  if (popularity <= 60) return "60";
  if (popularity <= 70) return "70";
  if (popularity <= 80) return "80";
  if (popularity <= 90) return "90"; 
  if (popularity <= 100) return "100"; 
}


function SpotifyTrends() {
  const [tracks, setTracks] = useState<Track[]>([]);
  const isMobile = window.innerWidth <= 960; // Common mobile cutoff

  useEffect(() => {
    async function downloadTopTracks() {
      const tracks = getTopTracks();
      setTracks(tracks);
    }
    downloadTopTracks();
  }, []);

  function trimName(name: string) {
    if (isMobile && name.length > 26) return `${name.slice(0, 26)}...`;
    if (name.length > 35) return `${name.slice(0, 35)}...`;
    return name;
  }

  return (
    <div className='spotify-trends-container'>
      <Carousel
        swipeable={isMobile}
        draggable={false}
        showDots={!isMobile}
        responsive={responsive}
        infinite={isMobile}
        autoPlay={isMobile ? true : false}
        autoPlaySpeed={10000}
        keyBoardControl={true}
        // customTransition="all .5"
        // transitionDuration={1000}
        containerClass="carousel-container"
        arrows={!isMobile}
        dotListClass="custom-dot-list-style"
        itemClass="carousel-item-padding-40-px"
      >
        {
          tracks.map((track, i) => {
            const imageSrc = track?.album?.images?.[0].url;
            const popularity = track.popularity;
            return (
              <Card className='spotify-track-card' key={`spotify-trend-${i}`}>
                  <Card.Img
                    variant='top'
                    src={imageSrc}
                    className="spotify-album-img"
                  />
                <Card.Header>
                  <a href={track?.external_urls?.spotify} target="_blank" rel="noopener noreferrer">
                    <b>{`#${i + 1} ${trimName(track.name)}`}</b>
                  </a>
                </Card.Header>
                <a href={track?.artists?.[0].external_urls.spotify} target="_blank" rel="noopener noreferrer">
                  <Card.Subtitle className='spotify-subtitle'>{track.artists?.[0].name}</Card.Subtitle>
                </a>
                <Card.Body>
                  <div className='spotify-body-contents'>
                    <>
                      <span className='spotify-trend-meter'>
                        Popularity
                        <div
                          className='spotify-trend-meter-bar'
                        >
                          <ProgressBar
                            animated
                            variant={determineTrendiness(popularity)}
                            now={popularity}
                          />
                        </div>
                      </span>
                      <audio controls={true} className="spotify-song-preview">
                        <source src={track.preview_url} type="audio/mpeg" />
                      </audio>
                    </>
                  </div>
                </Card.Body>
              </Card>
            )
          })
        }
      </Carousel>
    </div>
  );
}

export default SpotifyTrends;
