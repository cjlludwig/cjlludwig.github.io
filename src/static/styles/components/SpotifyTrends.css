.spotify-section {
  background-color: #588157;
  padding: 4rem 2rem;
  max-width: 1400px;
  margin: 0 auto;
  width: 100%;
}

.section-title {
  color: #DAD7CD;
  font-size: 2.5rem;
  margin-bottom: 3rem;
  text-align: center;
  text-shadow: 0 2px 4px rgba(0, 0, 0, 0.1);
}

.spotify-trends-container {
  margin: 2rem auto;
  padding: 1rem;
  max-width: 1400px;
  width: 100%;
}

.spotify-track-card {
  margin: 0.5rem;
  transition: transform 0.3s ease;
  height: 100%;
  background: rgba(52, 78, 65, 0.95);
  border: 1px solid rgba(218, 215, 205, 0.3);
  box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
}

.spotify-track-card:hover {
  transform: translateY(-5px);
  box-shadow: 0 8px 16px rgba(0, 0, 0, 0.2);
  border-color: rgba(218, 215, 205, 0.5);
  background: rgba(52, 78, 65, 0.98);
}

.spotify-album-img {
  height: 200px;
  object-fit: cover;
}

.spotify-subtitle {
  margin-top: 0.5rem;
  color: #DAD7CD;
}

.spotify-body-contents {
  display: flex;
  flex-direction: column;
  gap: 1rem;
  padding: 5px;
}

.spotify-trend-meter {
  display: flex;
  flex-direction: column;
  gap: 0.5rem;
}

.spotify-trend-meter-bar {
  height: 10px;
  border-radius: 5px;
  overflow: hidden;
}

.spotify-song-preview {
  width: 100%;
  margin-top: 0.5rem;
}

/* Analytics dashboard styles */
.analytics-dashboard {
  padding: 1rem 0;
  max-width: 1400px;
  margin: 0 auto;
}

.stats-container {
  display: flex;
  flex-direction: column;
  gap: 1.5rem;
}

.stat-item {
  text-align: center;
  padding: 1rem;
  border-radius: 8px;
  background-color: rgba(52, 78, 65, 0.95);
  border: 1px solid rgba(218, 215, 205, 0.3);
}

.stat-item h4 {
  font-size: 1rem;
  margin-bottom: 0.5rem;
  color: #DAD7CD;
}

.stat-value {
  font-size: 2rem;
  font-weight: bold;
  color: #A3B18A;
}

.most-popular-track {
  display: flex;
  gap: 1rem;
  align-items: center;
}

.most-popular-img {
  width: 100px;
  height: 100px;
  object-fit: cover;
  border-radius: 8px;
}

.most-popular-info {
  flex: 1;
}

.most-popular-info h4 {
  margin-bottom: 0.25rem;
  color: #E9EDC9;
}

.most-popular-info p {
  color: #DAD7CD;
  margin-bottom: 0.5rem;
}

/* Card styling to match project theme */
.spotify-trends-container .card {
  background: rgba(52, 78, 65, 0.95);
  border: 1px solid rgba(218, 215, 205, 0.3);
  box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
}

.spotify-trends-container .card-header {
  background-color: rgba(52, 78, 65, 0.98);
  border-bottom: 1px solid rgba(218, 215, 205, 0.3);
  color: #E9EDC9;
}

.spotify-trends-container .card-header a {
  color: #A3B18A;
  text-decoration: none;
}

.spotify-trends-container .card-header a:hover {
  color: #E9EDC9;
}

.spotify-trends-container .card-body {
  color: #DAD7CD;
}

.spotify-trends-container .nav-tabs .nav-link {
  color: #DAD7CD;
}

.spotify-trends-container .nav-tabs .nav-link.active {
  background-color: rgba(52, 78, 65, 0.95);
  color: #A3B18A;
  border-color: rgba(218, 215, 205, 0.3);
}

.spotify-trends-container .nav-tabs {
  border-bottom: 1px solid rgba(218, 215, 205, 0.3);
}

/* Responsive adjustments */
@media (max-width: 768px) {
  .most-popular-track {
    flex-direction: column;
    align-items: center;
    text-align: center;
  }
  
  .most-popular-img {
    width: 150px;
    height: 150px;
  }
  
  .spotify-trends-container {
    padding: 0.5rem;
  }
}
