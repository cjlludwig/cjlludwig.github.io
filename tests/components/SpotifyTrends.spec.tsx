import React from "react";
import renderer from "react-test-renderer";

import SpotifyTrends from "../../src/components/SpotifyTrends";

describe("<SpotifyTrends />", () => {
  test("renders correctly", () => {
    const tree = renderer
      .create(<SpotifyTrends />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});