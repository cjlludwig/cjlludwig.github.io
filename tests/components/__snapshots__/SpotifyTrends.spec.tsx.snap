// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SpotifyTrends /> renders correctly 1`] = `
<div
  className="spotify-trends-container"
>
  <div
    className="react-multi-carousel-list carousel-container "
    dir="ltr"
  >
    <ul
      className="react-multi-carousel-track "
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseMove={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      style={
        {
          "overflow": "unset",
          "transform": "translate3d(0px,0,0)",
          "transition": "none",
        }
      }
    />
  </div>
</div>
`;
